README.md
# NAME

Crypt::OpenSSL::SignCSR - Sign a Certificate Signing Request in XS.

# SYNOPSIS

```perl
use Crypt::OpenSSL::SignCSR;

my $signer = Crypt::OpenSSL::SignCSR->new(
                              $private_key_pem
                              {   # OPTIONAL
                                  days    => $days,   # Number of days for the certificate
                                  digest  => $digest, # Signature digest default (SHA256)
                                  format  => $format, # Output format "text" or "pem" (default)
                              });
my $cert   = $signer->sign(
                              $request, # CSR in PEM format
                          );

my $ret = $signer->set_days(3650);
my $ret = $signer->set_format("text");
my $ret = $signer->set_days("SHA512");

$cert   = $signer->sign( $request ); # CSR in PEM format
```

# DESCRIPTION

Allows a Certificate Signing Request (CSR) to be signed to create a
X509 PEM encoded Certificate.

# METHODS

## sign($csr)

Sign the provided CSR in PEM format.

Returns a signed certificate file in the specified format.

Arguments:

```
* $csr - a PEM format Certificate signing request.  You can create one with
Crypt::OpenSSL::PKCS10 or any other product capable of creating a signing request.
```

## set\_digest($digest)

Set the digest that should be used for signing the certificate.

Any openssl supported digest can be specified.  If the value provided is not
a valid it will set the openssl default.

Returns true (1) if successful and false (0) for a failure.

Arguments:

```
* $digest - the specified openssl supported digest (ex SHA1, SHA256, SHA384, SHA512)
```

## get\_digest()

Get the digest that is currently set.

Returns a string

## set\_format($format)

Set the format that should be used to output the the certificate.

Supported formats are "text" and "pem" (default).

Returns true (1) if successful and false (0) for a failure.

Arguments:

```
* $format - the specified output format ("pem", "text")
```

## get\_format()

Get the output format that is currently set.

Returns a string

## set\_days($days)

Set the number of days that the Certificate will be valid.  The days can
be set via the constructor or modified via set\_days()

Returns true (1) if successful and false (0) for a failure.

Arguments:

```
* $days - number of days that the certificate will be valid.
```

## get\_days()

Get the number of days that is currently set.

Returns the number of days or -1 for a failure.

# EXPORT

None by default.

# SEE ALSO

Crypt::OpenSSL::PKCS10 allows you to generate a Certificate Signing Request (CSR)

# AUTHOR

Timothy Legge, <timlegge@cpan.org>

# COPYRIGHT

Copyright (C) 2023 by Timothy Legge
Copyright 1995-2022 The OpenSSL Project Authors. All Rights Reserved.

I did not write any OpenSSL related code I simply copied and pasted
the work of the OpenSSL project's openssl code until I arrived at a XS
based module that could create a certificate from a Certificate Signing Request.

# LICENSE

Licensed under the Apache License 2.0 (the "License").  You may not use
this file except in compliance with the License.  You can obtain a copy
in the file LICENSE in the source distribution or at
https://www.openssl.org/source/license.html


Changes
Revision history for Crypt-OpenSSL-SignCSR.

{{$NEXT}}

0.14 -- Wed Jul 26 21:53:12 ADT 2023

    - 46298e7 Support openbsd's LibreSSL
    - f782397 v0.13

0.13 -- Tue Jul 25 23:43:27 ADT 2023

    - 63caf53 fix linux github actions
    - e6ce8ba A few more openssl differences and compiler issues
    - cfd7e85 Point to correct repo
    - f48933d use Crypt::OpenSSL:ConfiguredAPI to get the set API level
    - 0aa73aa v0.12

0.12 -- Thu Jul 20 23:41:31 ADT 2023

    - 150a974 Increment version for release
    - 61978d5 Fix incorrect case for PERL_VERSION
    - 9f7fafe Change tabs to spaces - no code changes
    - 10d3a65 Fix issue with more openssl versions
    - 9eccdd7 (tag: 0.11, origin/main) v0.11

0.11 -- Tue Jul 18 19:27:54 ADT 2023

    - daa5dd9 Increment version for release
    - 3a09798 Remove undocumented sigopts from sign()
    - 4295e6b Better regex for more openssl versions
    - a32ff18 Find out the version of OpenSSL
    - 2a21fa4 v0.10

0.10 -- Sat Jul 08 21:24:27 ADT 2023

    - 7272fa3 Increment the version for a release
    - 4b10acc Include all of 1.1 Openssl releases in older functions
    - b9d2ae4 v0.09

0.09 -- Sat Jul 08 15:26:53 ADT 2023

    [ Release 0.08 as production]
    - Some additional fixes for openssl versions since 0.07
    - 70debea Update version for release
    - 1e9fef5 v0.08

0.08 -- Fri Jul 07 23:58:11 ADT 2023

    - a6dc051 Fix up a few older openssl issues
    - 2e233f8 Increment repo versioni - in Makefile.PLwq
    - ee4d8c9 Increment repo version
    - 1676f26 (tag: 0.07) v0.07

0.07 -- Sun Jul 02 22:05:17 ADT 2023

    - 344b594 Some packaging fixes
    - 32fe761 Fix a pod issue and increment repo version
    - 40c37df (tag: 0.06) v0.06


0.06 -- Sat Jul 01 21:09:19 ADT 2023

    [Significant Changes since 0.05]

    Fix some remaining issues with openssl versions and
    a build issue on threaded perls which requires context

    [Detailed Logs]

    - ddfd2c2 Fix issue with threaded perl requiring context
    - 9b57abd More changes to fix warnings and errors for mac and windows
    - f063bee Increment version
    - 24b5db6 (tag: 0.05) v0.05

0.05 -- Fri Jun 30 17:26:04 ADT 2023

    [Significant Changes since 0.04]

    Breaking Changes: The sign() function now only accepts the
    Certificate Signing Request (CSR).  The other options have
    been moved to the constructor and functions are provided to
    update and obtain the values if required.

    - e7421ed Remove the unsupported macos 10.15 github action
    - 5996f1a Redesign of the interface and constructor
    - 642250e Add Crypt::OpenSSL::Guess as a dependency
    - f2e7378 Some additional OpenSSL change for version differences

0.04 -- Sun Jun 25 14:52:41 ADT 2023

    - Address issues compiling against older openssl versions

0.03 -- Tue Jun 20 10:27:40 ADT 2023

    - Fix a few packaging issues - no code changes

0.02 -- Mon Jun 19 23:57:58 ADT 2023

    - Early Official Release.  Same as version 0.01

0.01 -- Mon Jun 19 23:43:59 ADT 2023

	- Original Early Version - likely to change before official release



maint/Makefile_header.PL
use Config;
use File::Spec;
use Crypt::OpenSSL::Guess;
use Crypt::OpenSSL::ConfiguredAPI;

my %args;

my ($major, $minor, $patch) = openssl_version();
print "Installed OpenSSL: $major.$minor.$patch\n";

my $version = ($major * 10000) + ($minor * 100);
my $api = Crypt::OpenSSL::ConfiguredAPI::get_configured_api();
print "API version read $api\n";

my $compat = $api ne 0 ? $api : $version;
print "OPENSSL_COMPAT_API set to: $compat\n";

$args{INC} =  openssl_inc_paths();
$args{LIBS} = [openssl_lib_paths() . ' -lssl -lcrypto'];

my $cc_option_flags = " -DOPENSSL_API_COMPAT=$compat";

if ($Config::Config{cc} =~ /gcc/i) {
  $cc_option_flags .= $ENV{AUTHOR_TESTING} ? ' -Wall -Werror' : ' -Wall';
} else {
  $cc_option_flags .= '';
}

if ($Config{gccversion} =~ /llvm/i) {
  if ( $^O eq 'darwin' && (
    $Config{gccversion} =~ /LLVM 1[2-9].\d.[5-9]/ ||
    $Config{gccversion} =~ /LLVM 1[3-9]/ ))
  {
    $cc_option_flags .= ' -Wno-deprecated-declarations -Wno-compound-token-split-by-macro';
  } else {
    $cc_option_flags .= ' -Wno-deprecated-declarations';
  }

  if ($Config{gccversion} =~ /llvm 13/i) {
      $cc_option_flags .= ' -Wno-compound-token-split-by-macro';
  }

  if ($Config{PERL_VERSION} <= 20) {
    $cc_option_flags .= ' -Wno-error=pointer-bool-conversion';
  }
}

if ($Config{myuname} =~ /sunos|solaris/i) {
  $args{OPTIMIZE} = $cc_option_flags;
} else {
  $args{OPTIMIZE} = $cc_option_flags;
}


lib/Crypt/OpenSSL/SignCSR.pm
# ABSTRACT OpenSSL Self Sign a Certificate Signing Request in XS.
package Crypt::OpenSSL::SignCSR;

use 5.014;
use strict;
use warnings;

require Exporter;

our $VERSION  = "0.13";

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);

require XSLoader;
XSLoader::load('Crypt::OpenSSL::SignCSR', $VERSION);

1;
__END__

=head1 NAME

Crypt::OpenSSL::SignCSR - Sign a Certificate Signing Request in XS.

=head1 SYNOPSIS

  use Crypt::OpenSSL::SignCSR;

  my $signer = Crypt::OpenSSL::SignCSR->new(
                                $private_key_pem
                                {   # OPTIONAL
                                    days    => $days,   # Number of days for the certificate
                                    digest  => $digest, # Signature digest default (SHA256)
                                    format  => $format, # Output format "text" or "pem" (default)
                                });
  my $cert   = $signer->sign(
                                $request, # CSR in PEM format
                            );

  my $ret = $signer->set_days(3650);
  my $ret = $signer->set_format("text");
  my $ret = $signer->set_days("SHA512");

  $cert   = $signer->sign( $request ); # CSR in PEM format

=head1 DESCRIPTION

Allows a Certificate Signing Request (CSR) to be signed to create a
X509 PEM encoded Certificate.

=head1 METHODS

=head2 sign($csr)

Sign the provided CSR in PEM format.

Returns a signed certificate file in the specified format.

Arguments:

 * $csr - a PEM format Certificate signing request.  You can create one with
 Crypt::OpenSSL::PKCS10 or any other product capable of creating a signing request.

=head2 set_digest($digest)

Set the digest that should be used for signing the certificate.

Any openssl supported digest can be specified.  If the value provided is not
a valid it will set the openssl default.

Returns true (1) if successful and false (0) for a failure.

Arguments:

 * $digest - the specified openssl supported digest (ex SHA1, SHA256, SHA384, SHA512)

=head2 get_digest()

Get the digest that is currently set.

Returns a string

=head2 set_format($format)

Set the format that should be used to output the the certificate.

Supported formats are "text" and "pem" (default).

Returns true (1) if successful and false (0) for a failure.

Arguments:

 * $format - the specified output format ("pem", "text")

=head2 get_format()

Get the output format that is currently set.

Returns a string

=head2 set_days($days)

Set the number of days that the Certificate will be valid.  The days can
be set via the constructor or modified via set_days()

Returns true (1) if successful and false (0) for a failure.

Arguments:

 * $days - number of days that the certificate will be valid.

=head2 get_days()

Get the number of days that is currently set.

Returns the number of days or -1 for a failure.

=head1 EXPORT

None by default.

=head1 SEE ALSO

Crypt::OpenSSL::PKCS10 allows you to generate a Certificate Signing Request (CSR)

=head1 AUTHOR

Timothy Legge, E<lt>timlegge@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright (C) 2023 by Timothy Legge
Copyright 1995-2022 The OpenSSL Project Authors. All Rights Reserved.

I did not write any OpenSSL related code I simply copied and pasted
the work of the OpenSSL project's openssl code until I arrived at a XS
based module that could create a certificate from a Certificate Signing Request.

=head1 LICENSE

Licensed under the Apache License 2.0 (the "License").  You may not use
this file except in compliance with the License.  You can obtain a copy
in the file LICENSE in the source distribution or at
https://www.openssl.org/source/license.html

=cut


Makefile.PL
# This Makefile.PL for Crypt-OpenSSL-SignCSR was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.49.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.014;
use ExtUtils::MakeMaker;

use Config;
use File::Spec;
use Crypt::OpenSSL::Guess;
use Crypt::OpenSSL::ConfiguredAPI;

my %args;

my ($major, $minor, $patch) = openssl_version();
print "Installed OpenSSL: $major.$minor.$patch\n";

my $version = ($major * 10000) + ($minor * 100);
my $api = Crypt::OpenSSL::ConfiguredAPI::get_configured_api();
print "API version read $api\n";

my $compat = $api ne 0 ? $api : $version;
print "OPENSSL_COMPAT_API set to: $compat\n";

$args{INC} =  openssl_inc_paths();
$args{LIBS} = [openssl_lib_paths() . ' -lssl -lcrypto'];

my $cc_option_flags = " -DOPENSSL_API_COMPAT=$compat";

if ($Config::Config{cc} =~ /gcc/i) {
  $cc_option_flags .= $ENV{AUTHOR_TESTING} ? ' -Wall -Werror' : ' -Wall';
} else {
  $cc_option_flags .= '';
}

if ($Config{gccversion} =~ /llvm/i) {
  if ( $^O eq 'darwin' && (
    $Config{gccversion} =~ /LLVM 1[2-9].\d.[5-9]/ ||
    $Config{gccversion} =~ /LLVM 1[3-9]/ ))
  {
    $cc_option_flags .= ' -Wno-deprecated-declarations -Wno-compound-token-split-by-macro';
  } else {
    $cc_option_flags .= ' -Wno-deprecated-declarations';
  }

  if ($Config{gccversion} =~ /llvm 13/i) {
      $cc_option_flags .= ' -Wno-compound-token-split-by-macro';
  }

  if ($Config{PERL_VERSION} <= 20) {
    $cc_option_flags .= ' -Wno-error=pointer-bool-conversion';
  }
}

if ($Config{myuname} =~ /sunos|solaris/i) {
  $args{OPTIMIZE} = $cc_option_flags;
} else {
  $args{OPTIMIZE} = $cc_option_flags;
}

my %WriteMakefileArgs = (
  "ABSTRACT" => "Sign a Certificate Signing Request in XS.",
  "AUTHOR" => "Timothy Legge <timlegge\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Crypt::OpenSSL::ConfiguredAPI" => 0,
    "Crypt::OpenSSL::Guess" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "Crypt::OpenSSL::ConfiguredAPI" => 0,
    "Crypt::OpenSSL::Guess" => 0,
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Crypt-OpenSSL-SignCSR",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Crypt::OpenSSL::SignCSR",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "XSLoader" => 0
  },
  "TEST_REQUIRES" => {
    "Crypt::OpenSSL::Guess" => 0,
    "Crypt::OpenSSL::PKCS10" => "0.19",
    "Crypt::OpenSSL::RSA" => 0,
    "File::Slurper" => "0.012",
    "File::Temp" => 0,
    "File::Which" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.13",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    %args,
);

my %FallbackPrereqs = (
  "Crypt::OpenSSL::ConfiguredAPI" => 0,
  "Crypt::OpenSSL::Guess" => 0,
  "Crypt::OpenSSL::PKCS10" => "0.19",
  "Crypt::OpenSSL::RSA" => 0,
  "Exporter" => 0,
  "File::Slurper" => "0.012",
  "File::Temp" => 0,
  "File::Which" => 0,
  "Test::More" => "0.88",
  "XSLoader" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
